sdks/python/apache_beam/runners/interactive/caching/file_based_cache.py

```diff
@@ -0,0 +1,311 @@
+    return writer
+
+  def read(self, **kwargs):
+    """Returns an iterator over the contents of the cache.
+
+    Args:
+      kwargs (Dict[str, Any]): Arguments to be passed to the underlying reader
+          class.
+
+    Returns:
+      An iterator over the elements in the cache.
+    """
+    reader_kwargs = self._reader_kwargs.copy()
+    reader_kwargs.update(kwargs)
+    source = self.reader(**reader_kwargs)._source
+    range_tracker = source.get_range_tracker(None, None)
+    for element in source.read(range_tracker):
+      yield element
+
+  def write(self, elements):
+    """Writes a collection of elements into the cache.
+
+    Args:
+      elements (Iterable[Any]): A collection of elements to be written to cache.
+    """
+    self._num_writes += 1
+    if self.element_type is None:
+      # TODO(ostrokach): We might want to infer the element type from the first
+      # N elements, rather than reading the entire iterator.
+      elements = list(elements)
+      self.element_type = datatype_inference.infer_element_type(elements)
+    sink = self.writer()._sink
+    handle = sink.open(self._file_path_prefix)
+    try:
+      for element in elements:
+        sink.write_record(handle, element)
+    finally:
+      sink.close(handle)
+
+  def truncate(self):
+    """Removes all contents from the cache."""
+    FileSystems.delete(list(glob_files(self.file_pattern)))
+    FileSystems.create(self._file_path_prefix + ".empty").close()
+    self.element_type = None
+
+  def remove(self):
+    """Deletes the cache, including all underlying data.
+
+    The cache should not be used after this method is called.
+    """
+    self._finalizer()
+
+  @property
+  def removed(self):
+    return not self._finalizer.alive
+
+  def __del__(self):
+    self.remove()
+
+  @property
+  def file_pattern(self):
+    return self.location + '**'
+
+  @property
+  def _reader_kwargs(self):
+    reader_kwargs = {
+        k: v for k, v in self._writer_kwargs.items()
+        if k in self._reader_passthrough_arguments
+    }
+    if ("coder" in self._reader_passthrough_arguments and
+        "coder" not in reader_kwargs):
+      if self.default_coder is not None:
+        reader_kwargs["coder"] = self.default_coder
+      elif self.element_type is None:
+        reader_kwargs["coder"] = None
+      else:
+        reader_kwargs["coder"] = coders.registry.get_coder(self.element_type)
+    return reader_kwargs
+
+  @property
+  def _file_path_prefix(self):
+    return self.location + "-{:03d}".format(self._num_writes)
+
+  def _configure_finalizer(self, persist):
+    if persist:
+      return finalize(self, lambda: None)
+    else:
+      return finalize(
+          self, lambda pattern: FileSystems.delete(list(glob_files(pattern))),
+          self.file_pattern)
+
+
+class TextBasedCache(FileBasedCache):
+  """A ``PCollectionCache`` object which uses a text-based file format to store
+  the underlying data.
+  """
+  _reader_class = textio.ReadFromText
```

Comment:
It might be better to explicitly pass configurations like reader class, writer class, etc to the base classes constructor. Otherwise implementers would not be able to tell without looking into examples what needs to be implemented in subclasses.

Reply:
I had not considered doing it this way, but I do see that this approach has some advantages. I changed the `FileBasedCache` class to take `reader_class` and `writer_class` as arguments. Furthermore, I delegated the setting of default arguments to the child classes (e.g. using `functools.partial`), removing the need for the `_writer_kwargs` attribute.

However, in general I still think that the `PCollectionCache` interface should not include `reader_class` and `writer_class` as arguments. This is because, at least in the case of unbounded PCollections, the implementation of caches will be entirely backend-dependent (for example, `PubSubBasedCache` is likely to share little code with `KafkaBasedCache`). In those cases, having `reader_class` and `writer_class` as arguments is likely to cause confusion.